### PR TITLE
Fix Nullpointer for Gradle Parsing

### LINKED ISSUE #4793: Gradle updates fail with "undefined method `name' for nil:NilClass'"
**Package ecosystem**
gradle

**dependabot.yml content**
```
version: 2
updates:
  - package-ecosystem: gradle
    directory: /
    schedule:
      interval: daily
    open-pull-requests-limit: 10
    registries: "*"
registries:
  codeartifact:
    type: maven-repository
    url: https://redacted-private-repo.codeartifact.us-east-2.amazonaws.com/maven/releases
    username: aws
    password: ${{secrets.CODEARTIFACT_TOKEN}}
```
**Updated dependency**
The portion of the logs I posted below are for `org.slf4j:slf4j-api`but the same error happens with all dependencies for which a newer version is found.

**What you expected to see, versus what you actually saw**
**Expected:** PRs opened for dependencies that need updating.
**Actual:** ERRORs in the logs for each dependency that needs updating. No PRs opened.

**Images of the diff or a link to the PR, issue or logs**
```

updater | INFO <job_308987315> Checking if org.slf4j:slf4j-api 1.7.32 needs updating
  proxy | 2022/03/03 12:19:45 [016] GET https://repo.maven.apache.org:443/maven2/org/slf4j/slf4j-api/maven-metadata.xml
  proxy | 2022/03/03 12:19:45 [016] 200 https://repo.maven.apache.org:443/maven2/org/slf4j/slf4j-api/maven-metadata.xml
  proxy | 2022/03/03 12:19:45 [018] GET https://redacted-private-repo.codeartifact.us-east-2.amazonaws.com:443/maven/releases/org/slf4j/slf4j-api/maven-metadata.xml
  proxy | 2022/03/03 12:19:45 [018] * authenticating maven repository request (host: redacted-private-repo.codeartifact.us-east-2.amazonaws.com)
  proxy | 2022/03/03 12:19:45 [018] 404 https://redacted-private-repo.codeartifact.us-east-2.amazonaws.com:443/maven/releases/org/slf4j/slf4j-api/maven-metadata.xml
updater | INFO <job_308987315> Latest version is 1.7.36
updater | INFO <job_308987315> Requirements to unlock own
updater | INFO <job_308987315> Requirements update strategy 
updater | INFO <job_308987315> Updating org.slf4j:slf4j-api from 1.7.32 to 1.7.36
updater | I, [2022-03-03T12:19:45.574280 #7]  INFO -- sentry: ** [Raven] Sending event d6b3bee060384cb5b5423b0fbe2b6458 to Sentry
  proxy | 2022/03/03 12:19:45 [020] POST https://sentry.io:443/api/1451818/store/
  proxy | 2022/03/03 12:19:45 [020] 200 https://sentry.io:443/api/1451818/store/
updater | ERROR <job_308987315> Error processing org.slf4j:slf4j-api (NoMethodError)
updater | ERROR <job_308987315> undefined method `name' for nil:NilClass
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_parser/property_value_finder.rb:123:in `properties'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_parser/property_value_finder.rb:104:in `block in property_details'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_parser/property_value_finder.rb:103:in `each'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_parser/property_value_finder.rb:103:in `property_details'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_parser/property_value_finder.rb:111:in `property_value'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:156:in `block in evaluate_properties'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:154:in `scan'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:154:in `evaluate_properties'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:135:in `block in original_buildfile_declaration'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:134:in `each'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:134:in `find'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:134:in `original_buildfile_declaration'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:117:in `update_version_in_buildfile'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:73:in `block in update_buildfiles_for_dependency'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:61:in `each'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:61:in `update_buildfiles_for_dependency'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:27:in `block in updated_dependency_files'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:26:in `each'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/vendor/ruby/2.7.0/gems/dependabot-gradle-0.176.0/lib/dependabot/gradle/file_updater.rb:26:in `updated_dependency_files'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:678:in `generate_dependency_files_for'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:274:in `check_and_create_pull_request'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:82:in `check_and_create_pr_with_error_handling'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:56:in `block in run'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:56:in `each'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:56:in `run'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/lib/dependabot/update_files_job.rb:17:in `perform_job'
updater | ERROR <job_308987315> /home/dependabot/dependabot-updater/lib/dependabot/base_job.rb:35:in `run'
updater | ERROR <job_308987315> bin/update_files.rb:22:in `<main>'
```

### PR DESCRIPTION
**Pull request by sconvent**
This will fix issue #6384 

The problem was that Dependabot initially recognized correctly that there is no `build.gradle.kts` in the root folder, but later tried to access it anyways (See `top_level_build_file` and the comment above [here](https://github.com/dependabot/dependabot-core/compare/main...sconvent:dependabot-core:fix-nullpointer-gradle#diff-529cac3df5acb2177d1528ec5190f5c8a9e0c59d2d4bf094b5b014ebaab72cadR86))

This problem is solved with an additional `.compact()` statement, which will remove any `nil` entries in the list of files to be looked into.

Tested it by executing it on (an old state) of one of my repos with:
`bin/dry-run.rb gradle sconvent/dependabot-issue-demonstrator --commit addf67d1aa51dade24e47d1043abbd0baa54332a`

Works as expected (i.e. no Nullpointers)

It will probably also fix #4793, as it is a very similar issue. 

Happy to get some feedback as this is my first PR for Dependabot.

### CODE REVIEW

File: gradle/lib/dependabot/gradle/file_parser/property_value_finder.rb [hunk 1]
```diff
@@ -86,7 +86,7 @@ def property_details(property_name:, callsite_buildfile:)
           all_files = [callsite_buildfile, top_level_buildfile].concat(
             FileParser.find_includes(callsite_buildfile, dependency_files),
             FileParser.find_includes(top_level_buildfile, dependency_files)
-          )
+          ).compact()
```
Comment (jeffwidman) [inline gradle/lib/dependabot/gradle/file_parser/property_value_finder.rb:89]: I checked, and `FileParser.find_includes()` already protects against hitting NullPointer:
https://github.com/dependabot/dependabot-core/blob/5dc907e3148961c74ddc0b4494730f8e86118f79/gradle/lib/dependabot/gradle/file_parser.rb#L51

So this should be the only code change needed 👍

**Review by jeffwidman — COMMENTED** [review]
LGTM, however could you add a unit test for this scenario to prevent future regressions?

Odds of regressing on this exact problem are low, but the fact that we're hitting an error here shows we don't have any test coverage for this scenario of a missing top-level file... and odds of regressing in some other way for this scenario are high enough that we ought to get that test in place.

It should be straightforward to write by copying one of the existing tests, but if you run into questions / need help please let me know.

**Comment by jeffwidman** [comment]
Also, lint issue needs fixing, it's trivial 😁

**Comment by jeffwidman** [comment]
Actually there may be a unit test in https://github.com/dependabot/dependabot-core/pull/4256 I am on my phone so can't look in detail.

File: gradle/lib/dependabot/gradle/file_parser/property_value_finder.rb [hunk 2]
```diff
@@ -86,7 +86,7 @@ def property_details(property_name:, callsite_buildfile:)
           all_files = [callsite_buildfile, top_level_buildfile].concat(
             FileParser.find_includes(callsite_buildfile, dependency_files),
             FileParser.find_includes(top_level_buildfile, dependency_files)
```
Comment (jeffwidman) [inline gradle/lib/dependabot/gradle/file_parser/property_value_finder.rb:88]: I checked, and `FileParser.find_includes()` already protects against hitting NullPointer:
https://github.com/dependabot/dependabot-core/blob/5dc907e3148961c74ddc0b4494730f8e86118f79/gradle/lib/dependabot/gradle/file_parser.rb#L51

**Comment by sconvent** [comment]
Linting issue should be fixed now.

Sure, I'll add a testcase. Thanks for the hint where I can find one.

**Comment by jeffwidman** [comment]
Thanks. Feel free to squash your commits unless they're logically separate, or I'm also happy to do that at merge time. We aren't sticklers for single commit PR's esp if multiple discrete changes need to land in a single PR, but no need for a distinct commit just to fix a lint issue.

**Comment by sconvent** [comment]
Added a testcase (for each Groovy and Kotln DSL) and squashed everything into one commit.
The testcases are a bit different from the one you mentioned. @jeffwidman . But they failed without the changes in the actual code and now succeed.

**Review by jeffwidman — APPROVED** [review]
Thank you! Tests look reasonable to me.

**Pull request merged**